File: src/main/java/ankol/mod/merger/merger/MergerFactory.java
package ankol.mod.merger.merger;

import ankol.mod.merger.core.AbstractFileMerger;
import ankol.mod.merger.core.MergerContext;
import ankol.mod.merger.merger.scr.TechlandScrFileMerger;
import ankol.mod.merger.merger.xml.TechlandXmlFileMerger;
import cn.hutool.cache.Cache;
import cn.hutool.cache.CacheUtil;
import cn.hutool.core.io.FileUtil;
import cn.hutool.core.util.ReflectUtil;

import java.util.HashMap;
import java.util.Map;
import java.util.Optional;

/**
 * 合并器获取工厂，通过判断文件扩展名来获取对应支持的合并器
 *
 * @author Ankol
 */
public class MergerFactory {

    private static final Map<String, Class<? extends AbstractFileMerger>> mergerMap = new HashMap<>();
    /**
     * 实例化后的实例存储在缓存里，不去反复构造实例
     */
    private static final Cache<Class<?>, AbstractFileMerger> mergerCache = CacheUtil.newLFUCache(5);

    static {
        // 注册.scr格式的合并器
        registerMerger(TechlandScrFileMerger.class, ".scr", ".def", ".loot", ".phx", ".ppfx", ".ares", ".mpcloth");
        // 注册.xml文件的合并器
        registerMerger(TechlandXmlFileMerger.class, ".xml");
        /*Runtime.getRuntime().addShutdownHook(new Thread(() -> {
            DebugTool.printCacheUseRate("mergerCache", (AbstractCache) mergerCache);
        }));*/
    }

    /**
     * 注册一个合并器，并关联一个或多个文件扩展名。
     *
     * @param merger     合并器实例。
     * @param extensions 要关联的文件扩展名（例如 ".txt", ".xml"）。
     */
    private static void registerMerger(Class<? extends AbstractFileMerger> merger, String... extensions) {
        for (String ext : extensions) {
            mergerMap.put(ext.toLowerCase(), merger);
        }
    }

    /**
     * 根据文件名获取对应的合并器。
     *
     * @param fileName 文件名（包含扩展名）
     * @return 一个包含合并器实例的 {@link Optional}；如果找不到合适的合并器，则为空。
     */
    public static Optional<AbstractFileMerger> getMerger(String fileName, MergerContext context) {
        String extension = "." + FileUtil.extName(fileName);
        Class<? extends AbstractFileMerger> aClass = mergerMap.get(extension.toLowerCase());
        if (aClass == null) {
            return Optional.empty();
        } else {
            AbstractFileMerger fileMerger = mergerCache.get(aClass);
            if (fileMerger == null) {
                synchronized (MergerFactory.class) {
                    fileMerger = ReflectUtil.newInstance(aClass, context);
                    mergerCache.put(aClass, fileMerger);
                }
            } else {
                fileMerger.setContext(context);
            }
            return Optional.of(fileMerger);
        }
    }
}

File: src/main/java/ankol/mod/merger/merger/MergerFactory.kt
package ankol.mod.merger.merger

import ankol.mod.merger.domain.MergerContext
import ankol.mod.merger.merger.json.TechlandJsonFileMerger
import ankol.mod.merger.merger.scr.TechlandScrFileMerger
import ankol.mod.merger.merger.xml.TechlandXmlFileMerger
import java.util.*

/**
 * 合并器获取工厂，通过判断文件扩展名来获取对应支持的合并器
 *
 * @author Ankol
 */
object MergerFactory {
    /**
     * 合并器实例缓存
     */
    private val mergerMap: MutableMap<String, (MergerContext) -> AbstractFileMerger> = HashMap()

    init {
        //.scr格式的合并器
        registerMerger(
            ::TechlandScrFileMerger,
            ".scr", ".def", ".loot", ".phx", ".ppfx", ".ares", ".mpcloth", ".gpufx", ".chs", ".scd"
        )
        //.xml文件的合并器
        registerMerger(::TechlandXmlFileMerger, ".xml")
        //json格式合并器(.model也是json格式，但由于是模型文件，不提供支持合并)
        registerMerger(::TechlandJsonFileMerger, ".gui", ".json")
    }

    /**
     * 注册一个合并器，并关联一个或多个文件扩展名。
     *
     * @param mergerFactory 合并器创建函数。
     * @param extensions 要关联的文件扩展名（例如 .scr, .xml）。
     */
    private fun registerMerger(mergerFactory: (MergerContext) -> AbstractFileMerger, vararg extensions: String) {
        for (ext in extensions) {
            mergerMap[ext.lowercase(Locale.getDefault())] = mergerFactory
        }
    }

    /**
     * 根据文件名获取对应的合并器。
     *
     * @param fileName 文件名（包含扩展名）
     * @return 一个包含合并器实例的 [Optional]；如果找不到合适的合并器，则为空。
     */
    fun getMerger(fileName: String, context: MergerContext): AbstractFileMerger? {
        val extension = "." + fileName.substringAfterLast(".")
        val mergerFactory = mergerMap[extension.lowercase(Locale.getDefault())]
        return mergerFactory?.invoke(context)
    }
}
